src/ImageSharp/Formats/Tiff/Ifd/DirectoryReader.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using SixLabors.ImageSharp.Metadata.Profiles.Exif;
+
+namespace SixLabors.ImageSharp.Formats.Tiff
+{
+    /// <summary>
+    /// The TIFF IFD reader class.
+    /// </summary>
+    internal class DirectoryReader
+    {
+        private readonly TiffStream stream;
+
+        private readonly EntryReader tagReader;
+
+        private uint nextIfdOffset;
+
+        public DirectoryReader(TiffStream stream)
+        {
+            this.stream = stream;
+            this.tagReader = new EntryReader(stream);
+        }
+
+        public IEnumerable<IExifValue[]> Read()
+        {
+            if (this.ReadHeader())
+            {
+                return this.ReadIfds();
+            }
+
+            return null;
+        }
+
+        private bool ReadHeader()
+        {
+            ushort magic = this.stream.ReadUInt16();
+            if (magic != TiffConstants.HeaderMagicNumber)
+            {
+                throw new ImageFormatException("Invalid TIFF header magic number: " + magic);
```

Comment:
These exceptions should be static to allow inlining. See the other format thow helpers

Reply:
Done - https://github.com/SixLabors/ImageSharp/pull/1400